crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,755 @@
+    }
+
+ end:
+    OSSL_CRMF_POPO_free(crm->popo);
+    crm->popo = pp;
+
+    return 1;
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_CREATE_POPO, ERR_R_MALLOC_FAILURE);
+ err:
+    OSSL_CRMF_POPO_free(pp);
+    return 0;
+}
+
+/* returns 0 for equal, -1 for a < b or error on a, 1 for a > b or error on b */
+static int X509_PUBKEY_cmp(X509_PUBKEY *a, X509_PUBKEY *b)
+{
+    X509_ALGOR *algA = NULL, *algB = NULL;
+    int res = 0;
+
+    if (a == b)
+        return 0;
+    if (a == NULL
+        || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a) || algA == NULL)
+        return -1;
+    if (b == NULL
+        || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b) || algB == NULL)
+        return 1;
+    if ((res = X509_ALGOR_cmp(algA, algB)) != 0)
+        return res;
+    return EVP_PKEY_cmp(X509_PUBKEY_get0(a), X509_PUBKEY_get0(b));
+}
+
+/* verifies the Proof-of-Possession of the request with the given rid in reqs */
+int OSSL_CRMF_MSGS_verify_popo(const OSSL_CRMF_MSGS *reqs,
+                               int rid, int acceptRAVerified)
+{
+    OSSL_CRMF_MSG *req = NULL;
+    X509_PUBKEY *pubkey = NULL;
+    OSSL_CRMF_POPOSIGNINGKEY *sig = NULL;
+
+    if (reqs == NULL
+        || (req = sk_OSSL_CRMF_MSG_value(reqs, rid)) == NULL
+        || req->popo == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_MSGS_VERIFY_POPO,
+                CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    switch (req->popo->type) {
+    case OSSL_CRMF_POPO_RAVERIFIED:
+        if (acceptRAVerified)
+            return 1;
+        break;
+    case OSSL_CRMF_POPO_SIGNATURE:
+        pubkey = req->certReq->certTemplate->publicKey;
+        sig = req->popo->value.signature;
+        if (sig->poposkInput != NULL) {
+            /*-
+             * According to RFC 4211: publicKey contains a copy of
+             * the public key from the certificate template. This MUST be
+             * exactly the same value as contained in the certificate template.
+             */
+            if (pubkey == NULL
+                || sig->poposkInput->publicKey == NULL
+                || X509_PUBKEY_cmp(pubkey, sig->poposkInput->publicKey)
+                || ASN1_item_verify(ASN1_ITEM_rptr(OSSL_CRMF_POPOSIGNINGKEYINPUT),
+                                    sig->algorithmIdentifier, sig->signature,
+                                    sig->poposkInput,
+                                    X509_PUBKEY_get0(pubkey)) < 1)
+                break;
+        } else {
+            if (pubkey == NULL
+                || req->certReq->certTemplate->subject == NULL
+                || ASN1_item_verify(ASN1_ITEM_rptr(OSSL_CRMF_CERTREQUEST),
+                                    sig->algorithmIdentifier, sig->signature,
+                                    req->certReq,
+                                    X509_PUBKEY_get0(pubkey)) < 1)
+                break;
+        }
+        return 1;
+    case OSSL_CRMF_POPO_KEYENC:
+#ifdef OSSL_CMP_CERTREP_NEW_IMPLEMENTS_POPO_KEYENC /* TODO */
```

Comment:
What is this?

Reply:
This code is just a reminder what to do if the 'encrypted cert' method is implemented in `OSSL_CMP_certrep_new()` (which is not part of this PR).
I've re-phrased it as a simpler TODO comment.